File: merge-k-sorted-lists/merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> q = new PriorityQueue<>((o1,o2)->{
            return o1.val-o2.val;
        });
        for(int i=0;i<lists.length;i++){
            if(lists[i]!=null)
            q.add(lists[i]);
        }
        ListNode ans = new ListNode(-1);
        ListNode curr = ans;
        while(!q.isEmpty()){
            ListNode head = q.poll();
            ans.next = head;
            ans = ans.next;
            if(head.next!=null)
            q.add(head.next);
        }
        return curr.next;
    }
    
}

File: 23-merge-k-sorted-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        res = ListNode(-1)
        ans = res
        hm = {}
        minheap = []
        i = 0
        n = len(lists)
        visit = set()
        while len(visit)<n:
            curr_head = lists[i]
            if curr_head:
                heapq.heappush(minheap,curr_head.val)
                lists[i] = curr_head.next
            else:
                visit.add(i)
            if i+1 == n and minheap:
                addNode = ListNode(heapq.heappop(minheap))
                res.next = addNode
                res = res.next
            i = (i+1)%n
        while minheap:
            addNode = ListNode(heapq.heappop(minheap))
            res.next = addNode
            res = res.next
        return ans.next


"""
minheap => minimum among the lot in nlogn
index that cyclens around the length of lists and pick the corrrect list for processing
takes the elemtn, adds to the minheap, moves to the next list. 
"""